Monika After Story/game/zz_apikeys.rpy

```diff
@@ -0,0 +1,556 @@
+                )
+                self._set_buttons_ok()
+
+        def _set_state_no_update(self):
+            """
+            Sets text and button flags for when no update is found
+            """
+            self._main_text = "Cert already up to date!"
+            self._hide_all_buttons()
+            self._show_ok = True
+            self._show_retry = True
+
+        def _set_state_updated(self):
+            """
+            Sets text and button flags for whe cert was updated
+            """
+            self._main_text = "Cert updated!"
+            self._set_buttons_ok()
+
+
+init -979 python:
+
+
+    def mas_getAPIKey(feature):
+        """
+        gets the API key for a feature
+
+        IN:
+            feature - the string name of the feature to lookup.
+
+        RETURNS: the api key, as a string. Will be null string if no key found
+        """
+        return store.mas_api_keys.Data.api_keys.get(feature, "")
+
+
+    def mas_hasAPIKey(feature):
+        """
+        Checks if a feature has an API key
+
+        IN:
+            feature - string name of the feature to check
+
+        RETURNS: true if the feature has an API key, false otherwise
+        """
+        return bool(mas_getAPIKey(feature))
+
+
+    def mas_registerAPIKey(feature, display_name, on_change=None):
+        """
+        Registers a feature that accepts an API key.
+        Features are NOT registered if they already exist.
+
+        Can run a function when the api key is changed. This function should
+        return a tuple:
+            [0] - True if the key is valid, False if not
+            [1] - optional error message to show
+        The return value is primarily for setting a key. If no return value
+        is provided, the key is assumed valid.
+        The API key is passed in as the first param. The key will be a null
+        string if the key is being cleared.
+
+        IN:
+            feature - the string name of the feature to lookup
+            display_name - the display name the feature should use on screen
+            on_change - function to run when the API key is changed.
+
+        RETURNS: True if the feature was added, False if not.
+        """
+        if store.mas_api_keys.feature_registered(feature):
+            # feature already registered
+            return False
+
+        store.mas_api_keys.register_feature(
+            feature,
+            (display_name, on_change)
+        )
+        return True
+
+
+init -980 python in mas_api_keys:
+    import json
+    import os
+    import pygame
+    import store
+    import store.mas_utils as mas_utils
+
+    import store.mas_api_keys_data as Data
+
+
+    def feature_registered(feature):
+        """
+        Checks if a feature is registered for API keys
+
+        IN:
+            feature - feature to check (string)
+
+        RETURNS: True if the feature is registered for API keys
+        """
+        return feature in Data.registered_api_keys
```

Comment:
yes, its an abstraction of the namespace so store vars ("data") are separated from the functions ("behavior') - like 2/3rds of a model-view-controller (MVC) pattern. I think its cleaner and it also prevents `global` spam, although in this case I never had to set a var. 

>too generic for something used widely
the generic name `Data` is the point, you don't really have to know where it came from, just that it contains the data for this store module.

>looks like you're accessing a static class fields
Thats exactly the feel I'm going for here, like a dataclass object in py3. I feel anyone with those questions could quickly find where `Data` is coming from and see its an alias for an imported store module. I also aimed for this to be short and simple as opposed to `mas_api_keys_data`, which is readable but unnecessarily verbose in the context of this store `mas_api_keys`.

Reply:
> you don't really have to know where it came from, just that it contains the data for this store module

If I don't know where it came from and the name doesn't describe it, I can't know what it can contain. I'd really prefer verbosity over just `Data.x`.

> like a dataclass object in py3

Even dataclasses have proper naming so you can see what data you're working with.

Storemodules are designed to work akin to modules/packages, and I think we should use them that way. So `import mas_api_keys_data` or `from mas_api_keys_data import (...)`. If you want a shortcut, fair, but it should be descriptive. I don't see how MVC pattern helps our code here. Like if in the future we'd get more `import module as Data` and would have multiple `Data` references in our code, it'd take more time to figure out what's where than the time you saved not typing module name.

Reply:
The idea is that there would only be 1 Data module for each store, so you wouldn't have conflicting data references. if it was just Data, you could assume it came from a store called `<store_name>_data`. But right now I'm streaming and multi is agreeing with you so I'll probably just move data over into the same store, and not do this pattern.

Reply:
wait would you be fine if it were named `DataContext`?

Reply:
You see, if it was a truly dataclass, I'd be fine with it, but when it's actually a renpy module, it's not as clear. Having that data in a separate module is cool, but the name is still kinda ambiguous to me. You could call it `APIKeysData` or `APIKeysDataCtx`, but that makes it look like some object/static class when it's not. Now it also has a lengthy name, you could as well use `mas_api_keys_data`.

What I want to say is I don't like both this approach and the chosen name for this situation. But if you really want to use it here, then *maybe* `APIKeysData` could work.

Reply:
alright, I concede and have combined the data module with the normal one.